discovery.go

```diff
@@ -0,0 +1,301 @@
+package pubsub
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	discimpl "github.com/libp2p/go-libp2p-discovery"
+)
+
+var (
+	// poll interval
+
+	// DiscoveryPollInitialDelay is how long the discovery system waits after it first starts before polling
+	DiscoveryPollInitialDelay = 0 * time.Millisecond
+	// DiscoveryPollInterval is approximately how long the discovery system waits in between checks for whether the
+	// more peers are needed for any topic
+	DiscoveryPollInterval = 1 * time.Second
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+
+	// connector handles connecting to new peers found via discovery
+	connector *discimpl.BackoffConnector
+}
+
+// MinTopicSize returns a function that checks if a router is ready for publishing based on the topic size.
+// The router ultimately decides the whether it is ready or not, the given size is just a suggestion.
+func MinTopicSize(size int) RouterReady {
+	return func(rt PubSubRouter, topic string) (bool, error) {
+		return rt.EnoughPeers(topic, size), nil
+	}
+}
+
+// newDiscover creates a new peer discovery mechanism
+func newDiscover() *discover {
+	return &discover{
+		advertising: make(map[string]context.CancelFunc),
+		discoverQ:   make(chan *discoverReq, 32),
+		ongoing:     make(map[string]struct{}),
+		done:        make(chan string),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) error {
+	if d.discovery == nil || p == nil {
+		return nil
+	}
+
+	d.p = p
+
+	minBackoff, maxBackoff := time.Second*10, time.Hour
+	rng := rand.New(rand.NewSource(rand.Int63()))
+	backoff := discimpl.NewExponentialBackoff(minBackoff, maxBackoff, discimpl.FullJitter, time.Second, 5.0, 0, rng)
+	conn, err := discimpl.NewBackoffConnector(p.host, 100, time.Second*10, backoff)
```

Comment:
@vyzo @Stebalien Any suggestions on numbers here or if I should expose `WithDiscoveryDialingBackoff` as a PubSub option?

Reply:
1. We _should_ expose it.
2. I'm not sure what this timeout is. Given the amount of time/effort it usually takes to dial, I'd go for a default 2 minute backoff.

Reply:
There's the timeout which is how long we'll try to dial someone before calling it quits, and there's the backoff which is how long we should wait before dialing at all. Timeout is fixed and I don't really see a good reason to change that now and the backoff follows some function (in this case exponential + some randomness).

The 10 second fixed timeout number was grabbed from https://github.com/libp2p/go-libp2p-pubsub-router/blob/b47a5e34c4bd854836440a59faddab595499592e/pubsub.go#L539 which is how we've been doing bootstrapping thus far. No problem changing this to 2 minutes though if you think that's more reasonable.

Will expose it, but how do the exponential numbers look for backoff? The current numbers for backoffs after attempt N look like:

1: 10 seconds
2: Range [10, 25] seconds
3: [10, 125] seconds
4: [10 seconds, ~10.4 minutes]
5: [10 seconds, ~52 minutes]
6+: [10 seconds, 1 hour]

Reply:
Done. We're going to try these numbers for now. They are externally configurable and we can also always change the defaults later.